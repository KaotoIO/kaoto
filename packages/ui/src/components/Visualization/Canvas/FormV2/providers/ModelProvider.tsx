import { createContext, FunctionComponent, PropsWithChildren, useMemo } from 'react';

export interface ModelContextValue {
  model: unknown;
  errors?: Record<string, string[]>;
  onPropertyChange: (propName: string, value: unknown) => void;
  disabled?: boolean;
}

export const ModelContext = createContext<ModelContextValue>({ model: {}, onPropertyChange: () => {} });

export const ModelContextProvider: FunctionComponent<PropsWithChildren<ModelContextValue>> = ({
  model,
  errors,
  onPropertyChange,
  disabled,
  children,
}) => {
  const value = useMemo(() => {
    return { model, errors, onPropertyChange, disabled };
  }, [model, errors, onPropertyChange, disabled]);

  return <ModelContext.Provider value={value}>{children}</ModelContext.Provider>;
};
