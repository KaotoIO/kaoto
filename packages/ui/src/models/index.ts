export * from './camel-catalog-index';
export * from './camel-components-catalog';
export * from './camel-processors-catalog';
export * from './catalog-kind';
export * from './flow';
export * from './kamelets-catalog';
export * from './schema';
