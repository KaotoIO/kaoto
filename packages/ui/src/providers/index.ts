export * from './catalog-schema-loader.provider';
export * from './entities.provider';
