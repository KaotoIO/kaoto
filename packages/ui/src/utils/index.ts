export * from './camel-case-to-space';
export * from './camel-uri-helper';
export * from './catalog-schema-loader';
export * from './create-camel-properties-sorter';
export * from './event-notifier';
export * from './get-array-property';
export * from './get-custom-schema-from-kamelet';
export * from './get-field-groups';
export * from './get-required-properties-schema';
export * from './get-serialized-model';
export * from './get-user-updated-properties-schema';
export * from './get-value';
export * from './get-viznodes-from-graph';
export * from './init-visible-flows';
export * from './is-datamapper';
export * from './is-defined';
export * from './is-enum-type';
export * from './is-to-processor';
export * from './is-xslt-component';
export * from './join-path';
export * from './node-icon-resolver';
export * from './pipe-custom-schema';
export * from './promise-timeout';
export * from './resolve-ref-if-needed';
export * from './set-value';
export * from './update-kamelet-from-custom-schema';
export * from './weight-schema-properties';
