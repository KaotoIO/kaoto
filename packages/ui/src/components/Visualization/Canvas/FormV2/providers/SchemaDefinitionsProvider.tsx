import { createContext, FunctionComponent, PropsWithChildren } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';

type SchemaDefinitionsContextValue = Record<string, KaotoSchemaDefinition['schema']>;

export const SchemaDefinitionsContext = createContext<SchemaDefinitionsContextValue>({});

export const SchemaDefinitionsProvider: FunctionComponent<
  PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>
> = ({ schema, children }) => {
  const definitions: SchemaDefinitionsContextValue = schema.definitions ?? {};

  return <SchemaDefinitionsContext.Provider value={definitions}>{children}</SchemaDefinitionsContext.Provider>;
};
