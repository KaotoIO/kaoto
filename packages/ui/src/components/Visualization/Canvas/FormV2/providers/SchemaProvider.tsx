import { createContext, FunctionComponent, PropsWithChildren, useContext, useMemo } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { resolveSchemaWithRef } from '../../../../../utils';
import { SchemaDefinitionsContext } from './SchemaDefinitionsProvider';

export interface SchemaContextValue {
  schema: KaotoSchemaDefinition['schema'];
  definitions: Record<string, KaotoSchemaDefinition['schema']>;
}

export const SchemaContext = createContext<SchemaContextValue>({ schema: {}, definitions: {} });

export const SchemaProvider: FunctionComponent<PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>> = ({
  schema,
  children,
}) => {
  const { definitions, omitFields } = useContext(SchemaDefinitionsContext);

  const value = useMemo(() => {
    const resolvedSchema = resolveSchemaWithRef(schema, definitions);

    if (Array.isArray(resolvedSchema.anyOf)) {
      resolvedSchema.anyOf = resolvedSchema.anyOf.map((anyOfSchema) => resolveSchemaWithRef(anyOfSchema, definitions));
    }
    if (Array.isArray(resolvedSchema.oneOf)) {
      resolvedSchema.oneOf = resolvedSchema.oneOf.map((oneOfSchema) => resolveSchemaWithRef(oneOfSchema, definitions));
    }

    omitFields.forEach((field) => {
      delete resolvedSchema.properties?.[field];
    });

    return { schema: resolvedSchema, definitions };
  }, [definitions, omitFields, schema]);

  return <SchemaContext.Provider value={value}>{children}</SchemaContext.Provider>;
};
