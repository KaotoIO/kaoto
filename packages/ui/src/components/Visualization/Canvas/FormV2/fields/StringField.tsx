import { Button, TextInputGroup, TextInputGroupMain, TextInputGroupUtilities } from '@patternfly/react-core';
import { TimesIcon } from '@patternfly/react-icons';
import { FunctionComponent, useContext } from 'react';
import { isDefined } from '../../../../../utils';
import { useFieldValue } from '../hooks/field-value';
import { SchemaContext } from '../providers/SchemaProvider';
import { FieldProps } from '../typings';
import { FieldWrapper } from './FieldWrapper';

export const StringField: FunctionComponent<FieldProps> = ({ propName, required, onRemove: onRemoveProps }) => {
  const { schema } = useContext(SchemaContext);
  const { value = '', errors, onChange, disabled } = useFieldValue<string>(propName);
  const lastPropName = propName.split('.').pop();
  const ariaLabel = isDefined(onRemoveProps) ? 'Remove' : `Clear ${lastPropName} field`;

  const onFieldChange = (_event: unknown, value: string) => {
    onChange(value);
  };

  const onRemove = () => {
    if (isDefined(onRemoveProps)) {
      onRemoveProps(propName);
      return;
    }

    /** Clear field by removing its value */
    onChange(undefined as unknown as string);
  };

  const id = `${propName}-popover`;

  return (
    <FieldWrapper
      propName={propName}
      required={required}
      title={schema.title}
      type="string"
      description={schema.description}
      defaultValue={schema.default?.toString()}
      errors={errors}
    >
      <TextInputGroup validated={errors ? 'error' : undefined} isDisabled={disabled}>
        <TextInputGroupMain
          type="text"
          role="textbox"
          className="pf-m-icon kaoto-form__string-field"
          id={propName}
          name={propName}
          placeholder={schema.default?.toString()}
          value={value}
          onChange={onFieldChange}
          aria-label={schema.title}
          aria-describedby={id}
        />

        <TextInputGroupUtilities>
          <Button
            variant="plain"
            data-testid={`${propName}__clear`}
            onClick={onRemove}
            aria-label={ariaLabel}
            title={ariaLabel}
            icon={<TimesIcon />}
          />
        </TextInputGroupUtilities>
      </TextInputGroup>
    </FieldWrapper>
  );
};
