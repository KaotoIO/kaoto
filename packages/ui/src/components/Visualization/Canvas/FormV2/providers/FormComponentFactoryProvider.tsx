import { createContext, FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { BooleanField } from '../fields/BooleanField';
import { DisabledField } from '../fields/DisabledField';
import { ObjectField } from '../fields/ObjectField/ObjectField';
import { OneOfField } from '../fields/OneOfField/OneOfField';
import { StringField } from '../fields/StringField';
import { PasswordField } from '../fields/PasswordField';
import { FieldProps } from '../typings';

type FormComponentFactoryContextValue = (schema: KaotoSchemaDefinition['schema']) => FunctionComponent<FieldProps>;

export const FormComponentFactoryContext = createContext<FormComponentFactoryContextValue | undefined>(undefined);

export const FormComponentFactoryProvider: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const factory = useCallback<FormComponentFactoryContextValue>((schema) => {
    if (schema.format === 'password') {
      return PasswordField;
    }
    switch (schema.type) {
      case 'string':
      case 'number':
      case 'integer':
        return StringField;
      case 'boolean':
        return BooleanField;
      case 'object':
        return ObjectField;
      case 'array':
        return () => <div>Array field</div>;
    }

    if (Array.isArray(schema.oneOf)) {
      return OneOfField;
    } else if (Array.isArray(schema.anyOf)) {
      return () => <div>AnyOf field</div>;
    }

    return DisabledField;
  }, []);

  return <FormComponentFactoryContext.Provider value={factory}>{children}</FormComponentFactoryContext.Provider>;
};
