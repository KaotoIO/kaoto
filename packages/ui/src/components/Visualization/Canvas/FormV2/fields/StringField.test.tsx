import { act, fireEvent, render } from '@testing-library/react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { ROOT_PATH } from '../../../../../utils';
import { ModelContextProvider } from '../providers/ModelProvider';
import { SchemaProvider } from '../providers/SchemaProvider';
import { StringField } from './StringField';

describe('StringField', () => {
  it('should render', () => {
    const { container } = render(
      <ModelContextProvider model="Value" onPropertyChange={jest.fn()}>
        <StringField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should set the appropriate placeholder', () => {
    const wrapper = render(
      <ModelContextProvider model={undefined} onPropertyChange={jest.fn()}>
        <SchemaProvider schema={{ type: 'string', default: 'Default Value' }}>
          <StringField propName={ROOT_PATH} />
        </SchemaProvider>
      </ModelContextProvider>,
    );

    const input = wrapper.getByRole('textbox');
    expect(input).toHaveAttribute('placeholder', 'Default Value');
  });

  describe('onChange', () => {
    const stringSchema: KaotoSchemaDefinition['schema'] = { type: 'string' };
    const numberSchema: KaotoSchemaDefinition['schema'] = { type: 'number' };
    const cases = [
      { initialValue: 'Value', newValue: 'New Value', expectedValue: 'New Value', schema: stringSchema },
      { initialValue: '', newValue: ' ', expectedValue: ' ', schema: stringSchema },
      { initialValue: '', newValue: '123', expectedValue: '123', schema: stringSchema },
      { initialValue: '', newValue: '123', expectedValue: 123, schema: numberSchema },
      { initialValue: '', newValue: '2.', expectedValue: '2.', schema: stringSchema },
      { initialValue: '', newValue: '2.', expectedValue: '2.', schema: numberSchema },
      { initialValue: '', newValue: '2.0', expectedValue: '2.0', schema: stringSchema },
      { initialValue: '', newValue: '2.0', expectedValue: 2, schema: numberSchema },
      { initialValue: '', newValue: '2.05', expectedValue: '2.05', schema: stringSchema },
      { initialValue: '', newValue: '2.05', expectedValue: 2.05, schema: numberSchema },
    ];

    it.each(cases)(
      'should emit `$expectedValue` when the user writes `$newValue`',
      ({ initialValue, newValue, expectedValue, schema }) => {
        const onPropertyChangeSpy = jest.fn();

        const wrapper = render(
          <ModelContextProvider model={initialValue} onPropertyChange={onPropertyChangeSpy}>
            <SchemaProvider schema={schema}>
              <StringField propName={ROOT_PATH} />
            </SchemaProvider>
          </ModelContextProvider>,
        );

        const input = wrapper.getByRole('textbox');
        act(() => {
          fireEvent.change(input, { target: { value: newValue } });
        });

        expect(onPropertyChangeSpy).toHaveBeenCalledTimes(1);
        expect(onPropertyChangeSpy).toHaveBeenCalledWith(ROOT_PATH, expectedValue);
      },
    );
  });

  it('should clear the input when using the clear button', async () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider model="Value" onPropertyChange={onPropertyChangeSpy}>
        <StringField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const fieldActions = wrapper.getByTestId(`${ROOT_PATH}__field-actions`);
    act(() => {
      fireEvent.click(fieldActions);
    });

    const clearButton = await wrapper.findByRole('menuitem', { name: /clear/i });
    act(() => {
      fireEvent.click(clearButton);
    });

    expect(onPropertyChangeSpy).toHaveBeenCalledTimes(1);
    expect(onPropertyChangeSpy).toHaveBeenCalledWith(ROOT_PATH, undefined);
  });

  it('should show errors if available for its property path', () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider
        model="Value"
        errors={{ [ROOT_PATH]: ['error message'] }}
        onPropertyChange={onPropertyChangeSpy}
      >
        <StringField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const errorMessage = wrapper.getByText('error message');
    expect(errorMessage).toBeInTheDocument();
  });

  it('wraps value with RAW when Raw button is clicked', async () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider model="Value" onPropertyChange={onPropertyChangeSpy}>
        <StringField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const fieldActions = wrapper.getByTestId(`${ROOT_PATH}__field-actions`);
    act(() => {
      fireEvent.click(fieldActions);
    });

    const clearButton = await wrapper.findByRole('menuitem', { name: /raw/i });
    act(() => {
      fireEvent.click(clearButton);
    });

    expect(onPropertyChangeSpy).toHaveBeenCalledTimes(1);
    expect(onPropertyChangeSpy).toHaveBeenCalledWith(ROOT_PATH, 'RAW(Value)');
  });

  it('unwraps value from RAW when already wrapped', async () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider model="RAW(Test Value)" onPropertyChange={onPropertyChangeSpy}>
        <StringField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const fieldActions = wrapper.getByTestId(`${ROOT_PATH}__field-actions`);
    act(() => {
      fireEvent.click(fieldActions);
    });

    const raw = await wrapper.findByRole('menuitem', { name: /raw/i });
    act(() => {
      fireEvent.click(raw);
    });

    expect(onPropertyChangeSpy).toHaveBeenCalledTimes(1);
    expect(onPropertyChangeSpy).toHaveBeenCalledWith(ROOT_PATH, 'Test Value');
  });
});
