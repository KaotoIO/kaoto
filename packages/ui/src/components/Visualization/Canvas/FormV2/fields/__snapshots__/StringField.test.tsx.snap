// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringField should render 1`] = `
<div>
  <div
    class="pf-v6-c-form__group"
    data-testid="#__field-wrapper"
  >
    <div
      class="pf-v6-c-form__group-label"
    >
      <label
        class="pf-v6-c-form__label"
        for="#"
      >
        <span
          class="pf-v6-c-form__label-text"
        >
          #
        </span>
      </label>
        
      <span
        class="pf-v6-c-form__group-label-help"
      >
        <div
          style="display: contents;"
        >
          <span
            aria-label="More info for # field"
            class="pf-v6-c-button pf-m-plain pf-m-no-padding"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF6/Button"
            data-ouia-safe="true"
            role="button"
            tabindex="0"
            type="button"
          >
            <span
              class="pf-v6-c-button__icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v6-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
                />
              </svg>
            </span>
          </span>
        </div>
      </span>
    </div>
    <div
      class="pf-v6-c-form__group-control"
    >
      <div
        class="pf-v6-c-text-input-group"
      >
        <div
          aria-describedby="#-popover"
          class="pf-v6-c-text-input-group__main pf-m-icon kaoto-form__string-field"
          id="#"
        >
          <span
            class="pf-v6-c-text-input-group__text"
          >
            <input
              aria-label="Type to filter"
              class="pf-v6-c-text-input-group__text-input"
              name="#"
              role="textbox"
              type="text"
              value="Value"
            />
          </span>
        </div>
        <div
          class="pf-v6-c-text-input-group__utilities"
        >
          <button
            aria-expanded="false"
            aria-label="#__field-actions"
            class="pf-v6-c-menu-toggle pf-m-plain"
            data-ouia-component-id="OUIA-Generated-MenuToggle-plain-1"
            data-ouia-component-type="PF6/MenuToggle"
            data-ouia-safe="true"
            data-testid="#__field-actions"
            type="button"
          >
            <span
              class="pf-v6-c-menu-toggle__icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v6-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 192 512"
                width="1em"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
